### PR TITLE
Refactor: arithmetic and data types

### LINKED ISSUE #1532: 🔧 [Refactor] Decay Precision, Calculation and Saving Strategies
## 🔧 Refactor ticket

Ich habe mal untersucht, wie wir unsere Balance speichern.
Wir speichern diese * 10000 als Integer. Also mit 4 Nachkommastellen. Ich frage mich warum genau 4?
D.h. nach jeder Transaktion wird die neue Balance mit dem draufgerechneten Decay gespeichert. Damit potenzieren wir mit jeder Transaktion den Rundungsfehler. Es scheint so, dass dieser zu ungunsten des Nutzers ist, allerdings halte ich das für überdenkenswert an diesem Verhalten etwas zu ändern.

Unsere Decay-Zahl `0.99999997802044727`  hat eine Präzision von 16, unser Processing eine Präzision von 4 (`*10000` )
Ich schlage eine eine Präzision von 20 vor, da das das Maximum von `toFixed()`  ist und somit eine technische Grenze darstellt, die sicher im gesamten Js Raum gilt (Informationslage ist eher dünn):  <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toFixed>
```
(1 * Math.pow(0.99999997802044727, 1)).toFixed(20).toString().split('.')[1]
'99999997802044726658'
```

Auf  diese Weise könnten wir einen Gradido für 14 Jahre korrekt darstellen (mit unserer aktuellen Decay Zahl):
```
(numberExponentToLarge(1 * Math.pow(0.99999997802044727, 1*60*60*24*365*14))).toString().split('.')[1].length
> 20
(numberExponentToLarge(1 * Math.pow(0.99999997802044727, 1*60*60*24*365*15))).toString().split('.')[1].length
> 21
```
numberExponentToLarge: <https://codereview.stackexchange.com/questions/243164/convert-exponential-e-notation-numbers-to-decimals>

In der Datenbank könnten wir es wie folgend speichern:
```
`balance' DECIMAL(36,20)
```

Zuvor, war die Balance als BigInt(20) definiert, inklusive der 4 Nachkommastellen; sprich 16 Vorkommastellen. Daher ` 20+16 = 36`.
Damit sind wir auch in dem Bereich von MAX_SAFE_INTEGER seitens JS, welche 16 Stellen hat: <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER>

Die maximale Menge an Gradido für einen Account wäre dann `2^53 -1`  oder `9,007,199,254,740,991`  oder `nine quadrillion seven trillion one hundred ninety-nine billion two hundred fifty-four million seven hundred forty thousand nine hundred ninety-one` (Quelle: <https://stackoverflow.com/questions/307179/what-is-javascripts-highest-integer-value-that-a-number-can-go-to-without-losin>). Zusätzlich 20 Neunen nach dem Komma.

Die Geldmenge ist pro Nutzer-Account bei 17826 Gradido stabil. Über Kredite kann dies natürlich um ein Vielfaches überstiegen werden. Die AUF- und Community-Konten sind aber die, welche deutlich größere Mengen an Kapital erreichen können, da diese auf der gleichen stabilen Geldmenge mit der Anzahl der Nutzer skalieren:
Eine Community mit 10 Milliarden Nutzern, für welche jeweils 17,826 Gradido auf einem AUF- oder Community-Konto liegen können abgebildet werden:
`9,007,199,254,740,991  > 10,000,000,000*17,826` 
Die maximale Anzahl an Mitgliedern einer Community würde dann bei 500 Milliarden liegen
`9.007.199.254.740.991/17,826 = 505,284,374,214.12494`

Ich schlage daher vor, eine genauere Decay-Zahl mit Präzision von 20 zu ermitteln. Ich meine zu wissen, dass du ein Programm dafür geschrieben hast.
Darüber hinaus würde ich gerne sowohl die Speicherung als auch die Berechnung anpassen.

Nachtrag: Das sind rein theoretische Überlegungen und es gibt sicher noch unbedachte technische Hindernisse, die die Präzision oder maximale Menge an abbildbaren Gradido einschränken könnten

### PR DESCRIPTION
**Pull request by ulfgebhardt**
## 🍰 Pullrequest

This PR refactors the data types to `DECIMAL(40,20)` and corrects our transaction data.
Furthermore it optimizes the graphql types and unifies things on that end.

### Issues

- fixes #1532
- requires #1531

### Todo

- [X] None

### CODE REVIEW

File: backend/src/graphql/resolver/TransactionResolver.ts [hunk 1]
```diff
@@ -120,142 +76,70 @@ export class TransactionResolver {
...
-          decayTransaction.balance = roundCeilFrom4(balance)
-          decayTransaction.decayDuration = decay.decayDuration
-          decayTransaction.decayStart = decay.decayStart
-          decayTransaction.decayEnd = decay.decayEnd
-          transactions.push(decayTransaction)
-        }
+    // find involved users
+    let involvedUserIds: number[] = []
+    userTransactions.forEach((transaction: dbTransaction) => {
+      involvedUserIds.push(transaction.userId)
```
Comment (Mogge) [inline backend/src/graphql/resolver/TransactionResolver.ts:83]: Is this always me?

File: backend/src/graphql/resolver/TransactionResolver.ts [hunk 2]
```diff
@@ -120,142 +76,70 @@ export class TransactionResolver {
...
+      if (transaction.linkedUserId) {
```
Comment (Mogge) [inline backend/src/graphql/resolver/TransactionResolver.ts:84]:
```suggestion
      if (transaction.linkedUserId && !involvedUserIds.includes(transaction.linkedUserId)) {
```

File: backend/src/graphql/resolver/TransactionResolver.ts [hunk 3]
```diff
@@ -120,142 +76,70 @@ export class TransactionResolver {
...
+        involvedUserIds.push(transaction.linkedUserId)
       }
+    })
+    // remove duplicates
+    involvedUserIds = involvedUserIds.filter((value, index, self) => self.indexOf(value) === index)
```
Comment (Mogge) [inline backend/src/graphql/resolver/TransactionResolver.ts:89]: not needed anymore

File: backend/src/graphql/resolver/TransactionResolver.ts [hunk 4]
```diff
@@ -120,142 +76,70 @@ export class TransactionResolver {
...
+    const transactions: Transaction[] = []
 
-      if (order === Order.DESC) {
-        transactions.reverse()
-      }
+    // decay transaction
+    if (currentPage === 1 && order === Order.DESC) {
+      const now = new Date()
+      transactions.push(
+        virtualDecayTransaction(lastTransaction.balance, lastTransaction.balanceDate, now, self),
```
Comment (Mogge) [inline backend/src/graphql/resolver/TransactionResolver.ts:105]: lastTransaction my be null for new users

File: backend/src/graphql/resolver/TransactionResolver.ts [hunk 5]
```diff
@@ -120,142 +76,70 @@ export class TransactionResolver {
...
+      )
     }
 
-    const transactionList = new TransactionList()
-    transactionList.count = userTransactionsCount
-    transactionList.transactions = transactions
+    // transactions
+    for (let i = 0; i < userTransactions.length; i++) {
```
Comment (Mogge) [inline backend/src/graphql/resolver/TransactionResolver.ts:110]: foreEach

File: backend/src/graphql/resolver/TransactionResolver.ts [hunk 6]
```diff
@@ -120,142 +76,70 @@ export class TransactionResolver {
...
+      balanceGDT = Number(resultGDTSum.data.sum) || 0
+    } catch (err: any) {
+      // eslint-disable-next-line no-console
+      console.log('Could not query GDT Server', err)
     }
 
-    return transactionList
+    // Construct Result
+    return new TransactionList(
+      lastTransaction.balance,
```
Comment (Mogge) [inline backend/src/graphql/resolver/TransactionResolver.ts:138]: Decay must be included here!

**Review by Mogge — CHANGES_REQUESTED** [review]

File: backend/src/graphql/model/Transaction.ts [hunk 7]
```diff
@@ -2,58 +2,71 @@
...
+    this.balanceDate = transaction.balanceDate
+    if (!transaction.decayStart) {
+      this.decay = new Decay(transaction.balance, new Decimal(0), null, null, null)
+    } else {
+      this.decay = new Decay(
+        transaction.balance,
+        transaction.decay,
+        transaction.decayStart,
+        transaction.balanceDate,
+        (transaction.balanceDate.getTime() - transaction.decayStart.getTime()) / 1000,
```
[the remark this answers is not in the document]
Comment (ulfgebhardt) [reply backend/src/graphql/model/Transaction.ts]: not possible, its not a string, using `Math.round`

File: backend/src/graphql/resolver/TransactionResolver.ts (hunk 1 above)
Comment (ulfgebhardt) [reply backend/src/graphql/resolver/TransactionResolver.ts]: yes

File: backend/src/graphql/resolver/TransactionResolver.ts (hunk 4 above)
Comment (ulfgebhardt) [reply backend/src/graphql/resolver/TransactionResolver.ts:116]: they will have zero transactions - cool?

**Review by Mogge — APPROVED** [review]

**Pull request merged**